### PR TITLE
Support rest pattern in destructuring

### LINKED ISSUE #545: Ellipsis destruction is not allowed

### PR DESCRIPTION
**Pull request by ark120202**
Resolves #545.

### CODE REVIEW

**Comment by tomblind** [comment]
One way ObjectRest could be improved is by passing a set-like object for usedProperties instead of an array that has to be manually searched:
```ts
const {a, b, ...rest} = {a: "A", b: "B", c: "C", d: "D"};
```
```lua
function __TS__ObjectRest(target, usedProperties)
    local result = {}
    for property in pairs(target) do
        if usedProperties[property] == nil then
            result[property] = target[property]
        end
    end
    return result
end

local ____ = {
    a = "A",
    b = "B",
    c = "C",
    d = "D",
}
a = ____.a
b = ____.b
rest = __TS__ObjectRest(____, {
    a = true,
    b = true,
    rest = true,
})
```

File: src/LuaTransformer.ts [hunk 1]
```diff
@@ -3362,6 +3388,18 @@ export class LuaTransformer {
 
                     result.push(assignmentStatement);
                     break;
+                case ts.SyntaxKind.SpreadElement:
+                    if (index !== node.elements.length - 1) break;
```
Comment (Perryvw) [inline src/LuaTransformer.ts]: Should we just ignore this or should this throw an error? Is this valid TS?

File: src/TSHelper.ts [hunk 2]
```diff
@@ -60,6 +60,14 @@ export function isAssignmentPattern(node: ts.Node): node is ts.AssignmentPattern
     return ts.isObjectLiteralExpression(node) || ts.isArrayLiteralExpression(node);
 }
 
+export function isDestructuringAssignment(node: ts.Node): node is ts.DestructuringAssignment {
```
Comment (Perryvw) [inline src/TSHelper.ts]: Typescript has this one too, any reason we're not using that?

File: src/LuaTransformer.ts [hunk 3]
```diff
@@ -2025,17 +2055,15 @@ export class LuaTransformer {
...
-
             const statements: tstl.Statement[] = [];
 
-            // For nested bindings and object bindings, fall back to transformBindingPattern
+            // For object, nested, omitted or rest bindings fall back to transformBindingPattern
             if (
                 ts.isObjectBindingPattern(statement.name) ||
                 statement.name.elements.some(
-                    elem => (!ts.isBindingElement(elem) || !ts.isIdentifier(elem.name)) && !ts.isOmittedExpression(elem)
+                    e => ts.isBindingElement(e) && (!ts.isIdentifier(e.name) || e.dotDotDotToken)
```
[the remark this answers is not in the document]
Comment (ark120202) [reply src/LuaTransformer.ts]: `BindingElement.name` has a `BindingName` type, which is a union of `Identifier` and `BindingPattern`, so `ts.isIdentifier(e.name)` is true for `const [foo] = ...` and false for `const [[foo]] = ...`.

File: src/LuaTransformer.ts (hunk 1 above)
Comment (ark120202) [reply src/LuaTransformer.ts]: It's invalid TS and that's exactly why we shouldn't throw there. TypeScript already provides a diagnostic, so there's no point to break compilation for user.

File: src/TSHelper.ts (hunk 2 above)
Comment (ark120202) [reply src/TSHelper.ts]: Are you sure about that? I couldn't find any functions that have a `is ts.DestructuringAssignment` type guard

Comment (Perryvw) [reply src/TSHelper.ts]: https://github.com/microsoft/TypeScript/blob/master/src/compiler/utilities.ts#L4023

File: src/LuaTransformer.ts (hunk 1 above)
Comment (Perryvw) [reply src/LuaTransformer.ts]: I agree.

File: src/TSHelper.ts (hunk 2 above)
Comment (ark120202) [reply src/TSHelper.ts]: It's in the block with `@internal` annotation, so can't be used directly

**Pull request merged**